### PR TITLE
raftstore: add APPLY_PENDING_ENTRIES_GAUGE metrics

### LINKED ISSUE #7686: UCP: add metrics for channels of peer fsm and apply fsm
## Description

We need a way to calculate message count and memory occupation of messages in FSM's mailbox.

You can add a method in [msg.rs](https://github.com/tikv/tikv/blob/master/components/raftstore/src/store/msg.rs) to return the memory size of msg. And mutate the metrics before it's enqueued and after it's dequeued.

You probably should use local metrics to reduce performance impact. You can refer [local_metrics](https://github.com/tikv/tikv/blob/master/components/raftstore/src/store/local_metrics.rs) to see how local metrics work.

## Difficulty

- Medium

## Score

- 605

## Mentor(s)

- @BusyJay 
- @zhangjinpeng1987

## Recommended Skills

- Rust

### PR DESCRIPTION
**Pull request by Rustin-Liu**
Signed-off-by: Rustin-liu <rustin.liu@gmail.com>
UCP #7686

### What problem does this PR solve?

Issue Number: close #7686 

Problem Summary:

Add message count.

### What is changed and how it works?

What's Changed:

Add FSM_RECEIVEED_MESSAGES_COUNTER metrics.

### Related changes
None

### Check List 

Tests 

None

Side effects

None

### Release note 
None

### CODE REVIEW

**Comment by Rustin-Liu** [comment]
@zhangjinpeng1987  @5kbpers  @BusyJay  PTAL~
Thanks!

**Comment by 5kbpers** [comment]
I guess it would be better to add this metric for each fsm in the batch system. What do you think? @BusyJay

**Comment by Rustin-Liu** [comment]
@zhangjinpeng1987@BusyJay ping~

**Comment by BusyJay** [comment]
@5kbpers Indeed.

**Comment by BusyJay** [comment]
@5kbpers Such implementation may have high impact on performance. So I think it's OK to just monitor a summary instead.

**Comment by BusyJay** [comment]
@Rustin-Liu Apply needs to be observed too.

**Comment by Rustin-Liu** [comment]
>  Apply needs to be observed too.
@BusyJay  Added. PTAL~

**Comment by Rustin-Liu** [comment]
@5kbpers  @BusyJay  ping~

**Comment by Rustin-Liu** [comment]
@5kbpers @BusyJay ping~

**Comment by BusyJay** [comment]
I think we have some misunderstanding. The original issue asks for messages in mailbox, which are pending messages rather than handled messages.

**Comment by Rustin-Liu** [comment]
> I think we have some misunderstanding. The original issue asks for messages in mailbox, which are pending messages rather than handled messages.

@BusyJay  How about now?

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -697,6 +697,18 @@ where
...
+    message_count: i64,
+}
+
+impl<E> Drop for YieldState<E>
+where
+    E: KvEngine,
+{
+    fn drop(&mut self) {
+        if self.message_count > 0 {
+            APPLY_PENDING_MSGS_COUNTER.inc_by(self.message_count);
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/apply.rs:709]: I think you can refer to how `APPLY_PENDING_BYTES_GAUGE` works. I think it's enough to mutate the counter in the same places.

Comment (Rustin-Liu) [reply components/raftstore/src/store/fsm/apply.rs:709]: I don't understand what you mean, can you provide more details?

I think it's difficult mutate message counter and entries gauge in the same places.

Comment (BusyJay) [reply components/raftstore/src/store/fsm/apply.rs:709]: What I mean is counting the messages in the same place as `APPLY_PENDING_BYTES_GAUGE`. So when constructing `Apply`, besides updating pending bytes, it also updating pending entries:
```
APPLY_PENDING_BYTES_GAUGE.add(entries_mem_size);
APPLY_PENDING_ENTRY_COUNTER.add(entries.len());
```

**Comment by Rustin-Liu** [comment]
@sre-bot  /test

File: components/raftstore/src/store/fsm/apply.rs (hunk 1 above)
Comment (Rustin-Liu) [reply components/raftstore/src/store/fsm/apply.rs]: Fixed.  PTAL~
But for now, I am still confused about why the entry length is used for this metric.

Thanks for you reviewing.

File: components/raftstore/src/store/fsm/apply.rs [hunk 2]
```diff
@@ -2300,6 +2300,7 @@ impl<S: Snapshot> Apply<S> {
         let entries_mem_size =
             (ENTRY_MEM_SIZE * entries.capacity()) as i64 + get_entries_mem_size(&entries);
         APPLY_PENDING_BYTES_GAUGE.add(entries_mem_size);
+        APPLY_PENDING_MSGS_COUNTER.inc_by(entries.len() as i64);
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/apply.rs:2303]: It should be a guage and also updated in drop method.

Comment (Rustin-Liu) [reply components/raftstore/src/store/fsm/apply.rs:2303]: Does the name `APPLY_PENDING_MSGS` suitable for that? I think it's wired.

Comment (BusyJay) [reply components/raftstore/src/store/fsm/apply.rs:2303]: It can be changed to `APPLY_PENDING_ENTRIES_GAUGE`.

**Comment by Rustin-Liu** [comment]
@sre-bot /bench

**Comment by Rustin-Liu** [comment]
@sre-bot /bench

**Comment by Rustin-Liu** [comment]
@sre-bot /bench

**Comment by Rustin-Liu** [comment]
@sre-bot /integration-common-test

**Comment by Rustin-Liu** [comment]
@sre-bot /run-all-tests

**Comment by Rustin-Liu** [comment]
> 
> ## Benchmark Report
> 
> > Run Sysbench Performance Test on VMs
> 
> ```diff
> @@                               Benchmark Diff                               @@
> ================================================================================
> tidb: 2c5565061706fa5a296c52098253020cd9bb0dd7
> --- tikv: 5755b3e67e910486fcd8ae76084e9979eec2bcd3
> +++ tikv: f2d8713fefb7603ca2060bc4e2e67a28eea94273
> pd: 92e29af8378f43b3555cf7c266ef020010825597
> ================================================================================
> oltp_update_index:
>     * QPS: 4275.44 ± 2.11% (std=77.92) delta: -5.56% (p=0.050)
>     * Latency p50: 29.95 ± 2.12% (std=0.54) delta: 5.91%
>     * Latency p99: 56.10 ± 5.03% (std=1.68) delta: 4.62%
>             
> oltp_insert:
>     * QPS: 7001.81 ± 0.14% (std=7.85) delta: -1.69% (p=0.069)
>     * Latency p50: 18.28 ± 0.15% (std=0.02) delta: 1.73%
>     * Latency p99: 31.69 ± 6.41% (std=1.43) delta: 2.85%
>             
> oltp_read_write:
>     * QPS: 14353.37 ± 0.92% (std=98.11) delta: 1.54% (p=0.016)
>     * Latency p50: 178.68 ± 0.94% (std=1.22) delta: -1.53%
>     * Latency p99: 335.93 ± 1.20% (std=2.84) delta: -1.20%
>             
> oltp_point_select:
>     * QPS: 39512.32 ± 3.66% (std=1028.91) delta: -2.10% (p=0.274)
>     * Latency p50: 3.24 ± 3.78% (std=0.09) delta: 2.21%
>     * Latency p99: 10.75 ± 2.72% (std=0.22) delta: 0.47%
>             
> oltp_update_non_index:
>     * QPS: 4734.80 ± 4.35% (std=173.84) delta: 0.94% (p=0.776)
>     * Latency p50: 27.07 ± 4.41% (std=1.00) delta: -0.98%
>     * Latency p99: 43.92 ± 2.41% (std=0.75) delta: 0.00%
>             
> ```

@BusyJay  FYI.

**Comment by BusyJay** [comment]
@Rustin-Liu The benchmark you reference shows a performance regression. I bench it multiple times to see if it's normal.

**Comment by Rustin-Liu** [comment]
@sre-bot /bench

**Comment by Rustin-Liu** [comment]
@sre-bot /bench

**Comment by Rustin-Liu** [comment]
@sre-bot /bench

**Comment by Rustin-Liu** [comment]
@BusyJay  We already have 6 results. Could you please take a look? Thanks!

**Comment by Rustin-Liu** [comment]
@sre-bot /bench

**Pull request merged**